db/post_migrate/20210524114545_remove_ghost_statuses.rb

```diff
@@ -0,0 +1,5 @@
+class RemoveGhostStatuses < ActiveRecord::Migration[6.1]
+  def change
+    Status.where(uri: Tombstone.select(:uri)).destroy_all
```

Comment:
This should instead be `delete_all` or re-define explicitly `Status` and the expected hooks, in order to make the migration script robust so that it still works if ran on newer versions of the Mastodon code (e.g., updating multiple releases at once).

I'm also wondering whether a `JOIN` wouldn't be better here.

Also please bump the schema version in `db/schema.rb` and `lib/mastodon/maintenance_cli.rb`.

Reply:
is `delete_all` will be fine?
I really don't know about that

Reply:
That's a good question… there are a few callbacks that are usually called when statuses are deleted, as far as I know, those are:
- `unlink_from_conversations`, not calling it could cause some minor issues when such a deleted status is a DM
- `decrement_counter_caches`, not calling it would cause statuses/replies/reblog counts to be slightly off

I think omitting them is fairly minor, but it might be better to implement them. However, doing that in a reliable way would be particularly tricky (being a post-migration script, this could run directly after the previous migration script, or after all of them have run…)

Reply:
Is `destroy_all` instead of `delete_all` could be problem?
I think `destroy_all` calls all related callback like normally received "Delete" AP action that should be no problem.

Reply:
It does, but the issue is that it calls the Ruby code that is checked out at the moment the migration runs. However, as this is a migration script, this is not a safe option: the Ruby code is always written for the latest version of the database schema, while the migration script is here, well, to migrate things. What makes it worse is that it's a post-deployment migration script, so it could run either according to its migration serial number, among other migration scripts, or it could run after every pre-deployment migration script, depending on `SKIP_POST_DEPLOYMENT_MIGRATIONS`.